### ISSUE TITLE
lldpd discarding frame from tap interfaces

### ISSUE THREAD

**Issue by cherif09** [vincentbernat/lldpd#411]
Here is our setup: we have two machines on which lldp is launched. Lldp frames are sent through tap interfaces(one on each machine). We can see lldp frames from tcpdump on each tap interface but in debug traces we can see that frames are discarded :

"2020-08-03T00:05:34 [ DBG/interfaces] receive PDU from ethernet device ethsei1tap1
2020-08-03T00:05:34 [ DBG/receive] discard frame received on ethsei1tap1
2020-08-03T00:05:34 [ DBG/event] schedule next cleanup
2020-08-03T00:05:34 [ DBG/event] next cleanup in 120 seconds"

When we use a physical interface , we receive frames successfully and they are well decoded:
"2020-08-03T00:13:26 [ DBG/receive] receive a frame on etha01
2020-08-03T00:13:26 [ DBG/interfaces] receive PDU from ethernet device etha01
2020-08-03T00:13:26 [ DBG/receive] decode received frame on etha01
2020-08-03T00:13:26 [ DBG/decode] decode a received frame on etha01"

Here are some specific information related to our test:

lldpd.conf:

configure system interface pattern ethsei1tap1
configure lldp portidsubtype ifname
configure lldp tx-interval 30
configure ports ethsei1tap1 lldp management-addresses-advertisements
configure ports ethsei1tap1 lldp status rx-and-tx
configure med fast-start enable

lldp is launched with this following command:
/usr/sbin/lldpd -dddd -l -O /tmp/lldpd.conf -I ethsei1tap1 -C !*

here is the lscpi | grep net info:
02:00.0 Ethernet controller: Broadcom Corporation Device b671 (rev 11)
02:00.1 Ethernet controller: Broadcom Corporation Device b671 (rev 11)
03:00.0 Ethernet controller: Intel Corporation I210 Gigabit Network Connection (rev 03)
ec:00.0 Ethernet controller: Intel Corporation Device 1896

We wonder if the lldp  supports tap interface . We also don't know why frames are discarded by the deamon.
Thank you for your help.
regards.

**Comment by vincentbernat** [comment]
If you don't get another message in the log just before the message about the discarded frame, it usually means that lldpd saw the frame it just sent. It's unclear if you see incoming frames with tcpdump on each tap interface.

```
tcpdump ethsei1tap1 -pni  -vv -X ether host 01:80:c2:00:00:0e
```

How are the two tap interfaces connected? If there is a bridge between them, bridges do not pass LLDP frames by default (and in this case, `tcpdump` should not be able to see them either).

**Comment by cherif09** [comment]
Thank you for your response,
here is the setup with connection of our tap interfaces

  Machine 1                           switch                                   switch                  Machine 2
tap_Interface  *--------------*port* ---------*qsfp*--------*port*-----------*tap_interface

About the tcpdump , we can see lldp incoming and outcoming frames from each side on tap interfaces. So we don't know why lldpd deamon discarded them.

**Comment by vincentbernat** [comment]
Can you provide me with the output of the `tcpdump` command and of `ip maddr list dev ethsei1tap1`

**Comment by cherif09** [comment]
00:43:57.134953 LLDP, length 263
        Chassis ID TLV (1), length 7
          Subtype MAC address (4): 78:f9:b4:ac:1c:ef
          0x0000:  0478 f9b4 ac1c ef
        Port ID TLV (2), length 7
          Subtype MAC address (3): 78:f9:b4:ac:1c:ff
          0x0000:  0378 f9b4 ac1c ff
        Time to Live TLV (3), length 2: TTL 120s
          0x0000:  0078
        System Name TLV (5), length 13: RightSide Fct
          0x0000:  5269 6768 7453 6964 6520 4663 74
        System Description TLV (6), length 141
          Poky (Yocto Project Reference Distro) 2.1.1 (krogoth) Linux 4.19.129-g6b2572b-nokia_sm_x86 #1 SMP PREEMPT Thu Jul 30 07:33:05 UTC 2020 x86_64
          0x0000:  506f 6b79 2028 596f 6374 6f20 5072 6f6a
          0x0010:  6563 7420 5265 6665 7265 6e63 6520 4469
          0x0020:  7374 726f 2920 322e 312e 3120 286b 726f
          0x0030:  676f 7468 2920 4c69 6e75 7820 342e 3139
          0x0040:  2e31 3239 2d67 3662 3235 3732 622d 6e6f
          0x0050:  6b69 615f 736d 5f78 3836 2023 3120 534d
          0x0060:  5020 5052 4545 4d50 5420 5468 7520 4a75
          0x0070:  6c20 3330 2030 373a 3333 3a30 3520 5554
          0x0080:  4320 3230 3230 2078 3836 5f36 34
        System Capabilities TLV (7), length 4
          System  Capabilities [Bridge, WLAN AP, Router, Station Only] (0x009c)
          Enabled Capabilities [Station Only] (0x0080)
          0x0000:  009c 0080
        Management Address TLV (8), length 12
          Management Address length 5, AFI IPv4 (1): 10.14.51.52
          Interface Index Interface Numbering (2): 2
          0x0000:  0501 0a0e 3334 0200 0000 0200
        Management Address TLV (8), length 24
          Management Address length 17, AFI IPv6 (2): fd5f:4268:2387:a881:40:43ff:fe80:1000
          Interface Index Interface Numbering (2): 23
          0x0000:  1102 fd5f 4268 2387 a881 0040 43ff fe80
          0x0010:  1000 0200 0000 1700
        Port Description TLV (4), length 11: ethsei1tap1
          0x0000:  6574 6873 6569 3174 6170 31
        Organization specific TLV (127), length 9: OUI IEEE 802.3 Private (0x00120f)
          Link aggregation Subtype (3)
            aggregation status [supported], aggregation port ID 0
          0x0000:  0012 0f03 0100 0000 00
        Organization specific TLV (127), length 9: OUI IEEE 802.3 Private (0x00120f)
          MAC/PHY configuration/status Subtype (1)
            autonegotiation [none] (0x00)
            PMD autoneg capability [unknown] (0x8000)
            MAU type 10BASET fdx (0x000b)
          0x0000:  0012 0f01 0080 0000 0b
        End TLV (0), length 0
        0x0000:  0207 0478 f9b4 ac1c ef04 0703 78f9 b4ac  ...x........x...
        0x0010:  1cff 0602 0078 0a0d 5269 6768 7453 6964  .....x..RightSid
        0x0020:  6520 4663 740c 8d50 6f6b 7920 2859 6f63  e.Fct..Poky.(Yoc
        0x0030:  746f 2050 726f 6a65 6374 2052 6566 6572  to.Project.Refer
        0x0040:  656e 6365 2044 6973 7472 6f29 2032 2e31  ence.Distro).2.1
        0x0050:  2e31 2028 6b72 6f67 6f74 6829 204c 696e  .1.(krogoth).Lin
        0x0060:  7578 2034 2e31 392e 3132 392d 6736 6232  ux.4.19.129-g6b2
        0x0070:  3537 3262 2d6e 6f6b 6961 5f73 6d5f 7838  572b-nokia_sm_x8
        0x0080:  3620 2331 2053 4d50 2050 5245 454d 5054  6.#1.SMP.PREEMPT
        0x0090:  2054 6875 204a 756c 2033 3020 3037 3a33  .Thu.Jul.30.07:3
        0x00a0:  333a 3035 2055 5443 2032 3032 3020 7838  3:05.UTC.2020.x8
        0x00b0:  365f 3634 0e04 009c 0080 100c 0501 0a0e  6_64............
        0x00c0:  3334 0200 0000 0200 1018 1102 fd5f 4268  34..........._Bh
        0x00d0:  2387 a881 0040 43ff fe80 1000 0200 0000  #....@C.........
        0x00e0:  1700 080b 6574 6873 6569 3174 6170 31fe  ....ethsei1tap1.
        0x00f0:  0900 120f 0301 0000 0000 fe09 0012 0f01  ................
        0x0100:  0080 0000 0b00 00                        .......
00:43:59.454586 LLDP, length 262

**Comment by cherif09** [comment]
root@fct-0a:~ >ip maddr list dev ethsei1tap1
42:     ethsei1tap1
        link  33:33:00:00:00:01
        link  01:00:5e:00:00:01
        link  01:80:c2:00:00:0e static
        link  01:80:c2:00:00:03 static
        link  01:80:c2:00:00:00 static
        inet  224.0.0.1
        inet6 ff02::1
        inet6 ff01::1

**Comment by vincentbernat** [comment]
From what I understand, the LLDP packet you have from tcpdump is the one generated by lldpd, not one that is received on the TAP interface (it says "Port Description TLV (4), length 11: ethsei1tap1").

**Comment by cherif09** [comment]
we have the same tap interface name in both machine , the machine 1 name is the leftSide and the machine 2 is the rightSide, we receive both frames :
01:33:05.557524 LLDP, length 262
        Chassis ID TLV (1), length 7
          Subtype Local (7): fct-0a
          0x0000:  0766 6374 2d30 61
        Port ID TLV (2), length 7
          Subtype MAC address (3): 78:f9:b4:ae:ff:e7
          0x0000:  0378 f9b4 aeff e7
        Time to Live TLV (3), length 2: TTL 120s
          0x0000:  0078
        System Name TLV (5), length 12: LeftSide Fct
          0x0000:  4c65 6674 5369 6465 2046 6374
        System Description TLV (6), length 141
          Poky (Yocto Project Reference Distro) 2.1.1 (krogoth) Linux 4.19.129-g6b2572b-nokia_sm_x86 #1 SMP PREEMPT Thu Jul 30 07:33:05 UTC 2020 x86_64
          0x0000:  506f 6b79 2028 596f 6374 6f20 5072 6f6a
          0x0010:  6563 7420 5265 6665 7265 6e63 6520 4469
          0x0020:  7374 726f 2920 322e 312e 3120 286b 726f
          0x0030:  676f 7468 2920 4c69 6e75 7820 342e 3139
          0x0040:  2e31 3239 2d67 3662 3235 3732 622d 6e6f
          0x0050:  6b69 615f 736d 5f78 3836 2023 3120 534d
          0x0060:  5020 5052 4545 4d50 5420 5468 7520 4a75
          0x0070:  6c20 3330 2030 373a 3333 3a30 3520 5554
          0x0080:  4320 3230 3230 2078 3836 5f36 34
        System Capabilities TLV (7), length 4
          System  Capabilities [Bridge, WLAN AP, Router, Station Only] (0x009c)
          Enabled Capabilities [Station Only] (0x0080)
          0x0000:  009c 0080
        Management Address TLV (8), length 12
          Management Address length 5, AFI IPv4 (1): 10.14.51.51
          Interface Index Interface Numbering (2): 2
          0x0000:  0501 0a0e 3333 0200 0000 0200
        Management Address TLV (8), length 24
          Management Address length 17, AFI IPv6 (2): fd5f:4268:2387:a881:40:43ff:fe80:0
          Interface Index Interface Numbering (2): 23
          0x0000:  1102 fd5f 4268 2387 a881 0040 43ff fe80
          0x0010:  0000 0200 0000 1700
        Port Description TLV (4), length 11: ethsei1tap1
          0x0000:  6574 6873 6569 3174 6170 31
        Organization specific TLV (127), length 9: OUI IEEE 802.3 Private (0x00120f)
          Link aggregation Subtype (3)
            aggregation status [supported], aggregation port ID 0
          0x0000:  0012 0f03 0100 0000 00
        Organization specific TLV (127), length 9: OUI IEEE 802.3 Private (0x00120f)
          MAC/PHY configuration/status Subtype (1)
            autonegotiation [none] (0x00)
            PMD autoneg capability [unknown] (0x8000)
            MAU type 10BASET fdx (0x000b)
          0x0000:  0012 0f01 0080 0000 0b
        End TLV (0), length 0
        0x0000:  0207 0766 6374 2d30 6104 0703 78f9 b4ae  ...fct-0a...x...
        0x0010:  ffe7 0602 0078 0a0c 4c65 6674 5369 6465  .....x..LeftSide
        0x0020:  2046 6374 0c8d 506f 6b79 2028 596f 6374  .Fct..Poky.(Yoct
        0x0030:  6f20 5072 6f6a 6563 7420 5265 6665 7265  o.Project.Refere
        0x0040:  6e63 6520 4469 7374 726f 2920 322e 312e  nce.Distro).2.1.
        0x0050:  3120 286b 726f 676f 7468 2920 4c69 6e75  1.(krogoth).Linu
        0x0060:  7820 342e 3139 2e31 3239 2d67 3662 3235  x.4.19.129-g6b25
        0x0070:  3732 622d 6e6f 6b69 615f 736d 5f78 3836  72b-nokia_sm_x86
        0x0080:  2023 3120 534d 5020 5052 4545 4d50 5420  .#1.SMP.PREEMPT.
        0x0090:  5468 7520 4a75 6c20 3330 2030 373a 3333  Thu.Jul.30.07:33
        0x00a0:  3a30 3520 5554 4320 3230 3230 2078 3836  :05.UTC.2020.x86
        0x00b0:  5f36 340e 0400 9c00 8010 0c05 010a 0e33  _64............3
        0x00c0:  3302 0000 0002 0010 1811 02fd 5f42 6823  3..........._Bh#
        0x00d0:  87a8 8100 4043 fffe 8000 0002 0000 0017  ....@C..........
        0x00e0:  0008 0b65 7468 7365 6931 7461 7031 fe09  ...ethsei1tap1..
        0x00f0:  0012 0f03 0100 0000 00fe 0900 120f 0100  ................
        0x0100:  8000 000b 0000                           ......
01:33:11.556794 LLDP, length 266
        Chassis ID TLV (1), length 10
          Subtype Local (7): localhost
          0x0000:  076c 6f63 616c 686f 7374
        Port ID TLV (2), length 7
          Subtype MAC address (3): 78:f9:b4:ac:1c:ff
          0x0000:  0378 f9b4 ac1c ff
        Time to Live TLV (3), length 2: TTL 120s
          0x0000:  0078
        System Name TLV (5), length 13: RightSide Fct
          0x0000:  5269 6768 7453 6964 6520 4663 74
        System Description TLV (6), length 141
          Poky (Yocto Project Reference Distro) 2.1.1 (krogoth) Linux 4.19.129-g6b2572b-nokia_sm_x86 #1 SMP PREEMPT Thu Jul 30 07:33:05 UTC 2020 x86_64
          0x0000:  506f 6b79 2028 596f 6374 6f20 5072 6f6a
          0x0010:  6563 7420 5265 6665 7265 6e63 6520 4469
          0x0020:  7374 726f 2920 322e 312e 3120 286b 726f
          0x0030:  676f 7468 2920 4c69 6e75 7820 342e 3139
          0x0040:  2e31 3239 2d67 3662 3235 3732 622d 6e6f
          0x0050:  6b69 615f 736d 5f78 3836 2023 3120 534d
          0x0060:  5020 5052 4545 4d50 5420 5468 7520 4a75
          0x0070:  6c20 3330 2030 373a 3333 3a30 3520 5554
          0x0080:  4320 3230 3230 2078 3836 5f36 34
        System Capabilities TLV (7), length 4
          System  Capabilities [Bridge, WLAN AP, Router, Station Only] (0x009c)
          Enabled Capabilities [Station Only] (0x0080)
          0x0000:  009c 0080
        Management Address TLV (8), length 12
          Management Address length 5, AFI IPv4 (1): 10.14.51.52
          Interface Index Interface Numbering (2): 2
          0x0000:  0501 0a0e 3334 0200 0000 0200
        Management Address TLV (8), length 24
          Management Address length 17, AFI IPv6 (2): fd5f:4268:2387:a881:40:43ff:fe80:1000
          Interface Index Interface Numbering (2): 23
          0x0000:  1102 fd5f 4268 2387 a881 0040 43ff fe80
          0x0010:  1000 0200 0000 1700
        Port Description TLV (4), length 11: ethsei1tap1
          0x0000:  6574 6873 6569 3174 6170 31
        Organization specific TLV (127), length 9: OUI IEEE 802.3 Private (0x00120f)
          Link aggregation Subtype (3)
            aggregation status [supported], aggregation port ID 0
          0x0000:  0012 0f03 0100 0000 00
        Organization specific TLV (127), length 9: OUI IEEE 802.3 Private (0x00120f)
          MAC/PHY configuration/status Subtype (1)
            autonegotiation [none] (0x00)
            PMD autoneg capability [unknown] (0x8000)
            MAU type 10BASET fdx (0x000b)
          0x0000:  0012 0f01 0080 0000 0b
        End TLV (0), length 0
        0x0000:  020a 076c 6f63 616c 686f 7374 0407 0378  ...localhost...x
        0x0010:  f9b4 ac1c ff06 0200 780a 0d52 6967 6874  ........x..Right
        0x0020:  5369 6465 2046 6374 0c8d 506f 6b79 2028  Side.Fct..Poky.(
        0x0030:  596f 6374 6f20 5072 6f6a 6563 7420 5265  Yocto.Project.Re
        0x0040:  6665 7265 6e63 6520 4469 7374 726f 2920  ference.Distro).
        0x0050:  322e 312e 3120 286b 726f 676f 7468 2920  2.1.1.(krogoth).
        0x0060:  4c69 6e75 7820 342e 3139 2e31 3239 2d67  Linux.4.19.129-g
        0x0070:  3662 3235 3732 622d 6e6f 6b69 615f 736d  6b2572b-nokia_sm
        0x0080:  5f78 3836 2023 3120 534d 5020 5052 4545  _x86.#1.SMP.PREE
        0x0090:  4d50 5420 5468 7520 4a75 6c20 3330 2030  MPT.Thu.Jul.30.0
        0x00a0:  373a 3333 3a30 3520 5554 4320 3230 3230  7:33:05.UTC.2020
        0x00b0:  2078 3836 5f36 340e 0400 9c00 8010 0c05  .x86_64.........
        0x00c0:  010a 0e33 3402 0000 0002 0010 1811 02fd  ...34...........
        0x00d0:  5f42 6823 87a8 8100 4043 fffe 8010 0002  _Bh#....@C......
        0x00e0:  0000 0017 0008 0b65 7468 7365 6931 7461  .......ethsei1ta
        0x00f0:  7031 fe09 0012 0f03 0100 0000 00fe 0900  p1..............
        0x0100:  120f 0100 8000 000b 0000                 ..........

**Comment by vincentbernat** [comment]
OK. I don't understand how this could happen. Are you comfortable using gdb?

The code which is "rejecting" the packet is here:

```c
static int
iflinux_generic_recv(struct lldpd_hardware *hardware,
    int fd, char *buffer, size_t size,
    struct sockaddr_ll *from)
{
	int n, retry = 0;
	socklen_t fromlen;

retry:
	fromlen = sizeof(*from);
	memset(from, 0, fromlen);
	if ((n = recvfrom(fd, buffer, size, 0,
		    (struct sockaddr *)from,
		    &fromlen)) == -1) {
		if (errno == EAGAIN && retry == 0) {
			/* There may be an error queued in the socket. Clear it and retry. */
			levent_recv_error(fd, hardware->h_ifname);
			retry++;
			goto retry;
		}
		if (errno == ENETDOWN) {
			log_debug("interfaces", "error while receiving frame on %s (network down)",
			    hardware->h_ifname);
		} else {
			log_warn("interfaces", "error while receiving frame on %s (retry: %d)",
			    hardware->h_ifname, retry);
			hardware->h_rx_discarded_cnt++;
		}
		return -1;
	}
	if (from->sll_pkttype == PACKET_OUTGOING)
		return -1;
	return n;
}
```

Since you don't get either the message mentioned above, it means you are hitting the only case where the function can return `-1` and this is the check around the fact the captured packet is going out of the interface, inside of the other direction. You could recompile lldpd and modify `src/daemon/interfaces-linux.c` the `iflinux_generic_recv()` function:


```c
if (from->sll_pkttype == PACKET_OUTGOING) {
  log_debug("interfaces", "hello1");
  return -1;
}
log_debug("interfaces", "hello2");
return n;
```

Then, try again and check if you get `hello1` and/or `hello2` in the logs.

**Comment by cherif09** [comment]
2020-08-04T00:40:16 [ DBG/receive] receive a frame on ethsei1tap1
2020-08-04T00:40:16 [ DBG/interfaces] receive PDU from ethernet device ethsei1tap1
2020-08-04T00:40:16 [ DBG/interfaces] hello1
2020-08-04T00:40:16 [ DBG/receive] discard frame received on ethsei1tap1

**Comment by vincentbernat** [comment]
So, it means the packet is tagged as outgoing. Maybe the source MAC address is the same in both cases? You can see that with `tcpdump -e`.

**Comment by cherif09** [comment]
i've checked the src mac address but there are different but with the tcpdump -e i can see something that makes me suspicious :
>tcpdump -ei ethsei1tap1 ether host 01:80:c2:00:00:0e
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on ethsei1tap1, link-type EN10MB (Ethernet), capture size 262144 bytes
18:36:43.077229 78:f9:b4:ae:ff:e7 (oui Unknown) > 01:80:c2:00:00:0e (oui Unknown), ethertype LLDP (0x88cc), length 273: LLDP, length 259: localhost
18:36:51.556803 78:f9:b4:ac:1c:ff (oui Unknown) > 01:80:c2:00:00:0e (oui Unknown), ethertype LLDP (0x88cc), length 273: LLDP, length 259: localhost
18:37:13.081673 78:f9:b4:ae:ff:e7 (oui Unknown) > 01:80:c2:00:00:0e (oui Unknown), ethertype LLDP (0x88cc), length 273: LLDP, length 259: localhost
18:37:21.558365 78:f9:b4:ac:1c:ff (oui Unknown) > 01:80:c2:00:00:0e (oui Unknown), ethertype LLDP (0x88cc), length 273: LLDP, length 259: localhost

We can see a tag "oui Unknown" on each packet, i'm currently checking the reason of this tag.

**Comment by vincentbernat** [comment]
❦  7 septembre 2020 01:57 -07, cherif09:

> We can see a tag "oui Unknown" on each packet, i'm currently checking the reason of this tag.

It's not important. It's just the MAC address is not attached to a
vendor.

Try to temporarily comment the discard due to the `PACKET_OUTGOING` to
be sure that's the problem.

**Comment by cherif09** [comment]
It works when we comment the code related to the check of the sll_packetype.
Now we wonder why packets are considered as outgoing in our case and it seems like 
it's not related to lldp deamon but linux network stack.

**Comment by vincentbernat** [comment]
Which kernel version are you using? What is behind the tap interface? A qemu?

**Comment by cherif09** [comment]
we've finally found why packet are tagged outgoing. We have implemented a dispatcher in our switch which create an ACL rule to trap LLDP packets coming into specific switch ports.At this point packets are already considered as "incoming". Packets are copied into machine memeory and our dispatcher(which use a socket) sends them to tap interfaces on which lldp deamon is binding. The sending operation change the sll_pkttype to outgoing , that's why the lldp considers them as outgoing packets and discarded them. So to fix the issue , we should find a way to send packets without changing the sll_pkttype.

**Comment by cherif09** [comment]
we have fixed the issue by creating a second socket which will be the receiver and then use "sendto" api between the first socket of our lldp dispatcher and the new one. lldp packets reached that socket won't be tagged packet_outgoing. In the previous code , we used sendmsg api to send the lldp packet directly to the socket . In that way , packets will be tagged pkt_outgoing and then discarded by the lldp deamon. Thank you a lot for your help , we wouldn't find the RC without your help.
Regards.
